### PR TITLE
Rewrite Comparator System

### PR DESCRIPTION
**Pull request by APickledWalrus**
### Description

*NOTE: branch name is slightly inaccurate sorry - my original plans changed*
*NOTE: Not marking this as a breaking change as full backwards compatibility remains*

Welcome to the first of *many* upcoming codebase cleanup PRs!

The PR aims to rewrite and improve Skript's comparator system. By using stronger generics, better interface declarations, and complete documentation, the system is in a much stronger and better place. I've rewritten this system under the following new package: `org.skriptlang.skript.lang.comparators` (if you think someone else would be better, please let me know).

Full backwards compatibility has been kept by turning the old Comparator classes into wrapper classes for the new system.

Full list of changes:
- Broke up the comparator classes
- Improved documentation of all comparator classes
- Used clearer generics for easier to read code + stronger type enforcement
- Turned Comparator into a Functional Interface
- Added an overridable Comparator method for forbidding argument inversion (see InverseComparator)
- Improved the comparator search process to be clearer and to ensure it searches through all comparators between creating larger chains
- Maybe some other things I'm forgetting, idk

---
**Target Minecraft Versions:** Any
**Requirements:** N/A
**Related Issues:**
- https://github.com/SkriptLang/Skript/issues/4928

### CODE REVIEW

**Comment by APickledWalrus** [comment]
Looking for input on this:

Would it be a good idea to remove `QUICK_ACCESS_COMPARATORS` and instead store these specially resolved Comparators in the COMPARATORS list? New ComparatorInfos for them could be created if needed.

I'm not sure that the resolution of searching for `Pair<Class, Class>` is really any faster than just looping through the main list again to find an exact match from the Infos (in fact, then we would only be searching once, right?)

It would also mean that `getComparatorInfo()` returns specially resolved Comparators during runtime. I made ComparatorInfo public again for compatibility mainly, but I'm not sure it should be actually. To ensure that all possible Comparators can be resolved, addons/other parts of Skript should probably be calling methods that can trigger a Comparator search, right?

File: src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+	 * Please note that comparators may convert objects if necessary for comparisons.
+	 * @param firstType The first type for comparison.
+	 * @param secondType The second type for comparison.
+	 * @return A comparator capable of determine the {@link Relation} between two objects of 'firstType' and 'secondType'.
+	 * Will be null if no comparator capable of comparing two objects of 'firstType' and 'secondType' was found.
+	 */
+	@Nullable
+	@SuppressWarnings("unchecked")
+	public static <Type1, Type2> Comparator<Type1, Type2> getComparator(Class<Type1> firstType, Class<Type2> secondType) {
+		Pair<Class<?>, Class<?>> typePair = new Pair<>(firstType, secondType);
```
Comment (Moderocky) [inline src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:121]: You can speed this up and make it simpler.
Back up in ComparatorInfo you should add a hashCode based on (only) the two classes.
You can then search by the hash to find an exact match a lot quicker.

Comment (Moderocky) [reply src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:121]: Clearly, a Pair does the same thing eventually, but it's requiring the creation of a new thing when you can use an integer instead.

File: src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+		}
+
+		// Same class but no comparator
+		if (firstType != Object.class && secondType == firstType) {
+			return (Comparator<Type1, Type2>) EQUALS_COMPARATOR;
+		}
+
+		// Well, we tried!
+		return null;
```
Comment (Moderocky) [inline src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:238]: You could probably implement a 'default' comparator here (for the purposes of equals/sorting checks) that simply compares by equals and hashCode.

File: src/main/java/org/skriptlang/skript/lang/comparator/Relation.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		return false;
+	}
+
+	/**
+	 * Returns this Relation's string representation, which is similar to "equal to" or "greater than".
+	 */
+	@Override
+	public String toString() {
+		switch (this) {
+			case EQUAL:
```
Comment (Moderocky) [inline src/main/java/org/skriptlang/skript/lang/comparator/Relation.java:100]: This should be a parameter in the enum rather than a switch here.

**Review by Moderocky — CHANGES_REQUESTED** [review]
Looks sound for the most part, a few little pointers.

File: src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java (hunk 2 above)
Comment (TPGamesNL) [reply src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:238]: I don't think that's a good idea, Skript currently has an error if two objects cannot be compared, e.g. a number and a string. Returning a default comparator will break this

File: src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+		// Attempt converting one parameter but with reversed types
+		for (ComparatorInfo<?, ?> unknownInfo : COMPARATORS) {
+			if (!unknownInfo.comparator.supportsInversion()) // Unsupported for reversing types
+				continue;
+
+			ComparatorInfo<CType1, CType2> info = (ComparatorInfo<CType1, CType2>) unknownInfo;
+
+			if (info.secondType.isAssignableFrom(firstType)) { // Attempt to convert the second argument to the first comparator type
+				Converter<Type2, CType1> sc1 = Converters.getConverter(secondType, info.firstType);
```
Comment (TPGamesNL) [inline src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:194]: You mixed it up, secondType is already assignable from firstType. Same below

File: src/main/java/org/skriptlang/skript/lang/comparator/InverseComparator.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+final class InverseComparator<Type1, Type2> implements Comparator<Type1, Type2> {
+
+	private final Comparator<? super Type2, ? super Type1> comparator;
+
+	public InverseComparator(Comparator<? super Type2, ? super Type1> comparator) {
+		this.comparator = comparator;
+	}
+
+	@Override
+	public Relation compare(final Type1 o1, final Type2 o2) {
```
Comment (TPGamesNL) [inline src/main/java/org/skriptlang/skript/lang/comparator/InverseComparator.java:34]:
```suggestion
	public Relation compare(Type1 o1, Type2 o2) {
```

File: src/main/java/org/skriptlang/skript/lang/comparator/Relation.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	/**
+	 * Test whether this Relation is fulfilled if another is, i.e. if the parameter 'other' fulfils <code>X rel Y</code>,
+	 * then this Relation fulfils <code>X rel Y</code> as well.
+	 *
+	 * @param other The Relation to compare with.
+	 * @return Whether this Relation is part of the given Relation, e.g. <code>GREATER_OR_EQUAL.is(EQUAL)</code> returns true.
+	 */
+	public boolean is(Relation other) {
```
Comment (TPGamesNL) [inline src/main/java/org/skriptlang/skript/lang/comparator/Relation.java:63]: What do you think about naming this method `implies`?

File: src/main/java/ch/njol/skript/registrations/Converters.java [hunk 7]
```diff
@@ -287,9 +287,10 @@ public static boolean converterExists(final Class<?> from, final Class<?>... to)
 	 * @return the converter or null if none exist
 	 */
 	@Nullable
-	public static <F, T> Converter<? super F, ? extends T> getConverter(final Class<F> from, final Class<T> to) {
+	@SuppressWarnings("unchecked")
+	public static <F, T> Converter<F, T> getConverter(final Class<F> from, final Class<T> to) {
 		ConverterInfo<? super F, ? extends T> info = getConverterInfo(from, to);
-		return info != null ? info.converter : null;
+		return info != null ? (Converter<F, T>) info.converter : null;
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/registrations/Converters.java:293]: Why the changes to this method?

File: src/main/java/ch/njol/skript/classes/Comparator.java [hunk 8]
```diff
@@ -18,60 +18,52 @@
...
  * Used to compare two objects of a different or the same type.
- * 
- * @author Peter Güttinger
- * @param <T1> ,
- * @param <T2> the types to compare
- * @see Comparators#registerComparator(Class, Class, Comparator)
- * @see DefaultComparators
+ * @deprecated Use {@link org.skriptlang.skript.lang.comparator.Comparators}
  */
 public interface Comparator<T1, T2> {
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/classes/Comparator.java:25]: You also need to add the Deprecated annotation (not sure if it's needed explicitly on inner classes)

File: src/main/java/org/skriptlang/skript/lang/comparator/Relation.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package org.skriptlang.skript.lang.comparator;
+
+/**
+ * Represents a relation between two objects.
+ */
+public enum Relation {
```
Comment (TPGamesNL) [inline src/main/java/org/skriptlang/skript/lang/comparator/Relation.java:24]: This class needs a toString, see CondCompare#L141

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

File: src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+		Class<Type1> firstType,
+		Class<Type2> secondType,
+		Comparator<Type1, Type2> comparator
+	) {
+		Skript.checkAcceptRegistrations();
+
+		if (firstType == Object.class && secondType == Object.class)
+			throw new IllegalArgumentException("It is not possible to add a comparator between objects");
+
+		COMPARATORS.add(new ComparatorInfo<>(firstType, secondType, comparator));
```
[the remark this answers is not in the document]
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:89]: Probably, yeah

File: src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java (hunk 4 above)
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:194]: I don't think -

The Comparator's secondType is assignable from our firstType

Therefore, we are checking to determine if we can convert our secondType to the Comparator's firstType

Comment (TPGamesNL) [reply src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:194]: Yea I was a bit too quick there, I just misread it haha

File: src/main/java/ch/njol/skript/registrations/Converters.java (hunk 7 above)
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/registrations/Converters.java:293]: I originally did this to prevent having to cast some stuff, but I am going to revert it

I am working on a Converters cleanup as well, so any redundant casts for Comparators can be removed in the future if needed

File: src/main/java/org/skriptlang/skript/lang/comparator/Relation.java (hunk 6 above)
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/lang/comparator/Relation.java:63]: Yeah, I think that is a clearer name for the method - it shouldn't be hard for an addon to switch anyways

File: src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+	@SuppressWarnings("unchecked")
+	public static <Type1, Type2> Comparator<Type1, Type2> getComparator(Class<Type1> firstType, Class<Type2> secondType) {
+		if (Skript.isAcceptRegistrations())
+			throw new SkriptAPIException("Comparators cannot be retrieved until Skript has finished registrations.");
+
+		int hash = Objects.hash(firstType, secondType);
+
+		Comparator<Type1, Type2> comparator = (Comparator<Type1, Type2>) QUICK_ACCESS_COMPARATORS.get(hash);
+
+		if (comparator == null) {
```
Comment (TPGamesNL) [inline src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:140]: There are two scenarios where HashMap#get returns null:
- The map doesn't contain the key, in this case we should compute the result and update the cache
- The map does contain the key, but associated with a null value, in this case we shouldn't call getComparator_i, instead return null immediately (getComparator_i may return null if no comparator found, we should store this in the cache as well).

I think the easiest approach for this is to use computeIfAbsent

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

File: src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java [hunk 12]
```diff
@@ -132,17 +133,10 @@ public static <Type1, Type2> Relation compare(@Nullable Type1 first, @Nullable T
...
-
-		Comparator<Type1, Type2> comparator = (Comparator<Type1, Type2>) QUICK_ACCESS_COMPARATORS.get(hash);
-
-		if (comparator == null) {
-			comparator = getComparator_i(firstType, secondType);
-			QUICK_ACCESS_COMPARATORS.put(hash, comparator);
-		}
-
-		return comparator;
+		return (Comparator<Type1, Type2>) QUICK_ACCESS_COMPARATORS.computeIfAbsent(
```
Comment (TPGamesNL) [inline src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java:136]: Okay this was my bad but I looked at the description of `computeIfAbsent` and it also computes it when the key is mapped to null, so you can't use it 😅

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

**Review by Moderocky — APPROVED** [review]

**Comment by TheLimeGlass** [comment]
@APickledWalrus conflicts

**Comment by TheLimeGlass** [comment]
What's the point of quick converters? Shouldn't it be cleared? Why not use Guava cache and when it gets expired after X amount of time, then add it to the main converters collection.

**Comment by APickledWalrus** [comment]
> What's the point of quick converters? Shouldn't it be cleared? Why not use Guava cache and when it gets expired after X amount of time, then add it to the main converters collection.

The quick access system is *generally* just for caching comparators resolved during runtime, so that we aren't doing the chaining process each time.

**Review by TheLimeGlass — APPROVED** [review]
